connect/api/src/main/java/org/apache/kafka/connect/connector/ConnectorContext.java

```diff
@@ -30,4 +30,11 @@
      * added/removed) and the running Tasks will need to be modified.
      */
     void requestTaskReconfiguration();
+
+    /**
+     * Raise an unrecoverable exception to the Connect framework. This will cause the status of the
+     * connector to transition to FAILED.
+     * @param e Exception to be raised.
+     */
+    void raiseError(Exception e);
```

Comment:
This is interesting. Any reason not to handle this by throwing exceptions? In the clients, we have retriable/non-retriable exceptions to differentiate (although, admittedly when exceptions are coming from user code, it becomes less clear how to handle exceptions outside the exception hierarchy for Connect).

By the way, this type of issue was also filed as a JIRA separately from the status API: https://issues.apache.org/jira/browse/KAFKA-2479 It's possible this patch will also resolve that since it's basically required in order to figure out the status of the connector/task.